gpflow/monitor.py

```diff
@@ -0,0 +1,245 @@
+        """
+        It calls the 'run' function and sets the current step.
+
+        :param step: current step in the optimisation.
+        :param kwargs: additional key-word arguments that can be passed
+            to the `run` method of the task. This is in particular handy for
+            passing keyword argument to the callback of `ScalarToTensorBoard`.
+        """
+        self.current_step = tf.cast(step, tf.int64)
+        if step % self.period == 0:
+            self.run(**kwargs)
+
+    @abstractmethod
+    def run(self, **kwargs):
+        """
+        Implements the task to be executed on __call__.
+        The current step is available through `self.current_step`.
+
+        :param kwargs: keyword arguments available to the run method.
+        """
+        raise NotImplementedError
+
+
+class ToTensorBoard(MonitorTask):
+    def __init__(self, log_dir: str, period: int = 1):
+        """
+        :param log_dir: directory in which to store the tensorboard files.
+            Can be a nested. E.g. ./logs/my_run/
+        :param period: defines how often to run the task; it will execute every `period`th step.
+            For large values of `period` the task will be less frequently run.
+        """
+        super().__init__(period)
+        self.file_writer = tf.summary.create_file_writer(log_dir)
+
+    def __call__(self, step, **kwargs):
+        with self.file_writer.as_default():
+            super().__call__(step, **kwargs)
+        self.file_writer.flush()
+
+
+class ModelToTensorBoard(ToTensorBoard):
+    """
+    Monitoring task that creates a sensible TensorBoard for a model.
+
+    By default, it writes to the TensorBoard all model parameters that are scalars,
+    for parameter arrays (e.g. kernel.lengthscales) the values are sent to the TensorBoard
+    if the array has no more than `max_size` (default: 3) elements.
+    This behaviour can be adjusted using `max_size`.
+    """
+
+    def __init__(
+        self,
+        log_dir: str,
+        model: BayesianModel,
+        max_size: int = 3,
+        period: int = 1,
+        extra_keywords: List[str] = [],
+    ):
+        """
+        :param log_dir: directory in which to store the tensorboard files.
+            Can be a nested: for example, './logs/my_run/'.
+        :param model: model to be monitord.
+        :param max_size: maximum size of arrays (incl.) to store each
+            element of the array independently as a scalar in the TensorBoard.
+        :param period: defines how often to run the task; it will execute every `period`th step.
+            For large values of `period` the task will be less frequently run.
+        :param extra_keywords: pass additional keywords to be monitored.
+            If the parameter's name includes any of the extra keywords specified it
+            will be monitored as well. By default, parameters that match the `kernel` or
+            `likelihood` keyword are monitored.
```

Comment:
I'm wondering if it might be clearer if instead there was a `keywords` argument that simply had `["kernel", "likelihood"]` as its default value? rather than hardcoding those two?